[Feature] Structured Grading Instructions for Programming Exercises

fde312 opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I added multiple integration tests (Spring) related to the features (with a high test coverage)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I added multiple integration tests (Jest) related to the features (with a high test coverage)
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
Programming exercises did not feature the structured grading instructions.

### Description
<!-- Describe your changes in detail -->
- Replaced only problem statement with  `AssessmentInstructionsComponent`
- Added SGI functionality to programming exercises assessment
- Added link to solution repo for programming exercises in example solution section of `AssessmentInstructionsComponent`
- Added SGI view to `ProgrammingExerciseDetailsComponent`

### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->

1. Log in to Artemis
2. Navigate to Course Administration
3. Create programming exercise with due date and manual assessment
3.1 Now a toggle under grading instructions should be visible, toggle it and adapt the structured grading instructions
4. View programming exercise details and see if SGI are displayed
5. Participate in programming exercise
6. Assess exercise and check out the new instructions layout on the right side. It should contain Problem Statement, Example Solution, Grading Instruction
7. Add a inline feedback and drag and drop the structured grading instruction component to the inline feedback component
8. Inline feedback component should adapt accordingly 


### Test Coverage
<!-- Please add the test coverage for all changes files here. You can see this when executing the tests locally (see build.gradle and package.json) or when looking into the corresponding Bamboo build plan -->
* code-editor-tutor-assessment-inline-feedback.component.ts: 92%
* ProgrammingExerciseParticipationResource.java: 89%
* ProgrammingExerciseResource.java : 86%

### Screenshots
<!-- Add screenshots to demonstrate the changes in the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/2017/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->

SGI toggle:
<img width="1185" alt="image" src="https://user-images.githubusercontent.com/41108487/96338381-e39bbe00-108d-11eb-81a2-285ee72f1515.png">


Instructions on the right side:
<img width="1411" alt="image" src="https://user-images.githubusercontent.com/41108487/96333898-f81b8e80-106c-11eb-8c22-3256fb8dbdf1.png">

SGI before DnD:
<img width="1417" alt="image" src="https://user-images.githubusercontent.com/41108487/96333943-4466ce80-106d-11eb-9e33-4cc49f1bff2e.png">

After DnD:
<img width="1386" alt="image" src="https://user-images.githubusercontent.com/41108487/96333946-4df03680-106d-11eb-903b-c79ef1ce9883.png">

ProgrammingExerciseDetails view:
<img width="1184" alt="image" src="https://user-images.githubusercontent.com/41108487/96338213-bd295300-108c-11eb-9ae4-c5e30cff1169.png">

anditurdiu reviewed (commented):
The feature works as expected. I also tested the dragging instructors into the inline feedback and it worked like a charm. Good job! 
I left one minor comment to reduce unnecessary DB calls, but otherwise, it looks good to me.

anditurdiu commented on src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseParticipationResource.java:
```diff
@@ -93,10 +98,23 @@ public ProgrammingExerciseParticipationResource(ProgrammingExerciseParticipation
+
+        // Fetch template and solution participation into exercise of participation
+        ProgrammingExercise exercise = (ProgrammingExercise) participation.get().getExercise();
+        exercise = programmingExerciseService.findWithTemplateParticipationAndSolutionParticipationById(exercise.getId());
```
We only need the template and solution participation if the user is at least a teaching assistant or? 
In that case, it would be better if we move this into the if block in line `110`. This way we reduce unnecessary database calls. Same for the grading criteria in line `107`

fde312 replied:
jupe, you are right. I will add an else block where I add both calls 👍

FuchsDominik reviewed (approved):
Worked through the testing and everything works smooth. SGI are displayed like you said and dragging and dropping grading instructions also work, nice job!

schultek reviewed (approved):
This works as expected. Very cool feature. I just have a small comment on the code below.

tiffca commented:
Tested on test server 1. 
I couldn't make the inline comments work with the drag-and-drop of the structured grading criteria. I could only make it work with the add new feedback option. 
![grading-structured-criteria](https://user-images.githubusercontent.com/63906424/96467076-d7993300-122a-11eb-9f67-fd2e270974a7.gif)


In the add new feedback option, I'm able to add the structured grading criteria more than the defined limited number of times. 
<img src="https://user-images.githubusercontent.com/63906424/96466839-8e48e380-122a-11eb-8d9b-ea0abb81da3a.png" width="60%">

fde312 commented:
> Tested on test server 1.
> I couldn't make the inline comments work with the drag-and-drop of the structured grading criteria. I could only make it work with the add new feedback option.
> ![grading-structured-criteria](https://user-images.githubusercontent.com/63906424/96467076-d7993300-122a-11eb-9f67-fd2e270974a7.gif)
> 
> In the add new feedback option, I'm able to add the structured grading criteria more than the defined limited number of times.
> <img alt="" width="60%" src="https://user-images.githubusercontent.com/63906424/96466839-8e48e380-122a-11eb-8d9b-ea0abb81da3a.png">

@tiffca drag and drop directly into the code editor does not work. You need to hover first over the gutter (element with the line numbers), press on the plus button and then drag and drop the SGI into the **inline feedback component**. See screenshot "SGI before DnD:" and "SGI after DnD". 

I will take a look into the second issue. :)

fde312 commented on src/main/webapp/app/exercises/programming/manage/update/programming-exercise-update.component.ts:
```diff
@@ -59,6 +59,7 @@ export class ProgrammingExerciseUpdateComponent implements OnInit {
     existingCategories: ExerciseCategory[];
 
     public inProductionEnvironment: boolean;
+    checkedFlag = false;
```
Done in `97326be`

fde312 commented:
@tiffca I investigated your second issue:
- SGI allows to drag and drop the SGI elements as often as you want, there is no limit
- SGI takes only the amount of maxCountInScore into account when calculating the score
 -> example: SGI maxCountInScore = 2 with credits=1 and you drag and dropped the SGI element 5 times, then the score should be 2 even though 5 elements with credits 1 are displayed, because only 2 SGI elements are taken into account.
- Unrelated to this PR: When having a maxCountInScore = 0 and having credits != 0 then the feedback is taken into account in the calculation of the score as often as you drag and dropped it. @krusche @HanyaElhashemy is this an issue or not even a use case?

MaisiKoleni reviewed (changes requested):
When I cancel grading, the feedback element remains. The GIF below shows that when I start a new submission. It seems to happen only at the specific line, but for all files.

![qgi-for-progr](https://user-images.githubusercontent.com/11130248/96510292-953d1980-125d-11eb-961b-b58bfb5fc5a6.gif)

FuchsDominik reviewed (approved):
Tested this with @fde312 and adressed @tiffca remark like Francisco just stated. Except for the cornercase that the maximum usage of a grading instruction is equal 0, the calculation of the total score works fine even with a higher usage of a feedback than the max count.

FuchsDominik reviewed (commented):
Tested this with @fde312 and adressed @tiffca remark like Francisco just stated. Except for the cornercase that the maximum usage of a grading instruction is equal 0, the calculation of the total score works fine even with a higher usage of a feedback than the max count.
I will test it again and see if @MaisiKoleni  's problem occurs for me as well.

fde312 commented:
> When I cancel grading, the feedback element remains. The GIF below shows that when I start a new submission. It seems to happen only at the specific line, but for all files.
> 
> ![qgi-for-progr](https://user-images.githubusercontent.com/11130248/96510292-953d1980-125d-11eb-961b-b58bfb5fc5a6.gif)

@MaisiKoleni Good catch! Should be fixed now, can you retest? :)

tiffca commented:
thanks to the screen capture above I was able to make the in-line grading work. 

However the manual grade is not updating for me. 

![2222-score-not-changing](https://user-images.githubusercontent.com/63906424/96599382-265cd080-12f0-11eb-8092-facdf13ca2b3.gif)

kloessst reviewed (changes requested):
After saving an inline feedback, and opening the feedback box again, the box can't be closed anymore.

Otherwise everything else works as expected.

fde312 commented:
> thanks to the screen capture above I was able to make the in-line grading work.
> 
> However the manual grade is not updating for me. Expected is that there would be points updated under manual result, but it still shows 0/5.
> 
> ![2222-score-not-changing](https://user-images.githubusercontent.com/63906424/96599382-265cd080-12f0-11eb-8092-facdf13ca2b3.gif)

@tiffca This is the case, because it was not saved yet. In you screencast, you are still in the edit mode. After you clicked on saved, the feedback will be processed as manual feedback and then it will count into the calculation :) 

@kloessst You issue should also be fixed :)

kloessst commented on src/main/webapp/app/exercises/programming/manage/update/programming-exercise-update.component.ts:
```diff
@@ -59,6 +59,7 @@ export class ProgrammingExerciseUpdateComponent implements OnInit {
     existingCategories: ExerciseCategory[];
 
     public inProductionEnvironment: boolean;
+    checkedFlagForStructuredGradingInstructions = false;
```
For editing of programming exercises: Is is possible to check whether grading instructions have been created and then set this flag accordingly?

kloessst replied:
Nevermind, it seems like this was conscious decision when the structured grading instructions were created

kloessst reviewed (approved):
Code looks good to me. Everything works nicely and the structured grading criteria integrate nicely with your feature. Well done!

krusche reviewed (approved):
Code looks good to me after a quick review. I will merge this. In case of issues, @fde312 can also resolve those in a follow up PR.

HanyaElhashemy commented on src/main/webapp/app/exercises/modeling/assess/modeling-assessment-editor/modeling-assessment-editor.component.ts:
```diff
@@ -500,6 +502,6 @@ export class ModelingAssessmentEditorComponent implements OnInit {
      */
     calculateTotalScore() {
-        this.totalScore = (this.feedback || []).reduce((totalScore, feedback) => totalScore + feedback.credits!, 0);
+        this.totalScore = this.structuredGradingCriterionService.computeTotalScore(this.feedback);
```
Just a small note: the `maxCountInScore` feature is not integrated in Apollon yet so calculating the score using `structuredGradingCriterionService.computeTotalScore(this.feedback)` in modeling-assessment-editor will not change the score calculation (it will skip all maxCountInScore related conditions). Have you tested assessing modeling exercises after these changes? (everything should work as expected but just in case ^^')

HanyaElhashemy reviewed (commented):
Very cool to see the SGIs also used for programming exercises :) code wise everything looks good to me, well done!
(hope it was easy extending SGIs to support programming exercises ^^')
And to your question from above: maxCountInScore = 0 means that there is no maximum so the points of the instruction will be calculated in the totalscore as often as applied by the assessor
